cloud/store/migration.go

```diff
@@ -0,0 +1,178 @@
+// If the current node is not the leader, the function will return early without error.
+// If the current node is the leader, it will read the schema from schemaRepo and import it in the RAFT based
+// representation using st.Execute to ensure RAFT has log entries for the schema.
+// It returns an error if any happens during the migration process.
+func (st *Store) MigrateToRaft(schemaRepo schemaTypes.SchemaRepo) error {
+	log := st.log.With("action", "raft_migration")
+
+	// Fist let's ensure either a raft node is a leader, or this node is the leader.
+	ticker := time.NewTicker(500 * time.Millisecond)
+	// Using a for loop with ticker.C instead of select allows us to have an immediate "tick" at 0 instead of waiting
+	// the delay for the first tick
+	for ; true; <-ticker.C {
+		// Raft is shutting down, no leader found so far let's exit
+		if !st.open.Load() {
+			return nil
+		}
+
+		if st.IsLeader() || st.Leader() != "" {
+			log.Info("leader found", "leader", st.Leader())
+			break
+		}
+	}
+
+	// If we are not leader we abort the process, this ensure only one node will try to migrate from non raft to raft
+	// representation at the same time
+	if !st.IsLeader() {
+		log.Info("Node is not raft leader, skipping migration")
+		return nil
+	}
+
+	// Get the schema from the non-raft representation and load it from disk.
+	// If no schema is present then exit early as there is nothing to migration from
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancelFunc()
+	stateGetter, err := schemaRepo.Load(ctx)
+	if err != nil {
+		return fmt.Errorf("could not load non-raft schema to memory for migration: %w", err)
+	}
+	shardingState := stateGetter.GetShardingState()
+	schema := stateGetter.GetSchema()
+	if schema == nil || len(schema.Classes) <= 0 {
+		log.Info("no schema available or schema is empty in non-raft representation, nothing to migrate")
+		return nil
+	}
+
+	// Load current migration progress from the disk.
+	// If no progress was made so far, ensure the migrationProgress struct is initialized empty
+	migrationProgressFile := filepath.Join(st.raftDir, raftMigrationFileName)
+	migrationProgress := NewMigrationProgress()
+	migrationProgressFromDisk, err := os.ReadFile(migrationProgressFile)
+	recoveringMigration := false
+	// If an error happens here we don't want to abort the migration to have a smoother experience. Instead wipe the
+	// file and consider to migration took place.
+	if err != nil && !os.IsNotExist(err) {
+		log.Warn(fmt.Sprintf("could not read raft migration progress file: %s", err))
+	} else {
+		recoveringMigration = true
+		err = json.Unmarshal(migrationProgressFromDisk, migrationProgress)
+		if err != nil {
+			log.Warn("could not unmarshall raft migration progress file contents, migration will proceed and not consider any progress made")
+			// Ensure we reset the progress struct to a pristine state
+			migrationProgress = NewMigrationProgress()
```

Comment:
**thought (non-blocking):** is reassigning  needed ?

Reply:
I want to be sure that failing the json unmarshal doesn't leave any artifact in the struct. So re-assigning to ensure it's empty seems to be the safest way to do it.